crates/sui-indexer/src/models/watermarks.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::schema::watermarks::{self};
+use diesel::prelude::*;
+
+/// Represents a row in the `watermarks` table.
+#[derive(Queryable, Insertable, Default, QueryableByName)]
+#[diesel(table_name = watermarks, primary_key(entity))]
+pub struct StoredWatermark {
+    /// The table name of group of tables governed by this watermark, i.e `epochs`, `checkpoints`,
+    /// `transactions`.
+    pub entity: String,
+    /// Upper bound epoch range to enable per-entity epoch-level retention policy. Committer
+    /// advances this along with `high`.
+    pub epoch_hi: i64,
+    /// Lower bound epoch range to enable per-entity epoch-level retention policy. Pruner advances
+    /// this.
+    pub epoch_lo: i64,
+    pub checkpoint_hi: i64,
+    /// The inclusive high watermark that the committer advances.
+    pub hi: i64,
+    /// The inclusive low watermark that the pruner advances. Data before this watermark is
+    /// considered pruned.
+    pub lo: i64,
+    /// Pruner sets this, and uses this column to determine whether to prune or wait long enough
+    /// that all in-flight reads complete or timeout before it acts on an updated watermark.
+    pub timestamp_ms: i64,
+    /// Pruner updates this, and uses this when recovering from a crash to determine where to
+    /// continue pruning. Represents the latest watermark pruned, inclusive.
+    pub pruned_lo: Option<i64>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Ord, PartialOrd)]
+pub enum WatermarkEntity {
+    Checkpoints,
+    Events,
+    EvLookup,
+    // EventEmit,
+    // EventSenders,
+    // EventStruct,
+    ObjectsHistory,
+    Transactions,
+    TxLookup,
+    // TxAffectedAddresses,
+    // TxAffectedObjects,
+    // TxCalls,
+    // TxChangedObjects,
+    // TxDigests,
+    // TxInputObjects,
+    // TxKinds,
+    // TxRecipients,
+    // TxSenders,
+}
+
+pub struct Watermark {
+    pub entity: WatermarkEntity,
+    pub update: WatermarkUpdate,
+}
+
+pub enum WatermarkUpdate {
+    /// The subset of fields that the committer updates.
+    UpperBound { epoch_hi: u64, hi: u64 },
+    /// The subset of fields that the pruner updates.
+    LowerBound { epoch_lo: u64, lo: u64 },
+    /// Pruner logs its progress to this field. Uses this when recovering from a crash to determine
+    /// where to continue pruning. Represents the latest watermark pruned, inclusive.
+    LowerBoundTail { pruned_lo: u64 },
```

Comment:
Maybe it's just me but I find it quite error-prone that the `u64`s in this enum can mean different things (epoch, cp, tx) for different tables. I can imagine myself adding a new table but using the wrong sequence number for it. Would it make sense to use different variants for different sequence numbers? Like `LowerBoundTailCp`, `LowerBoundTailTx` so the type system can help reduce bugs here?